📖  Add code organization proposal (#13859)
kubernetes-sigs/cluster-api: docs/proposals/20260701-code-organization.md

---
title: Proposal Template
authors:
  - "@fabriziopandini"
reviewers:
  - "@sbueringer"
  - "@sivchari"
  - "@Karthik-K-N "
creation-date: 2026-07-01
last-updated: 2026-07-01
status: implementable
---

# Code organization

## Table of Contents

<!-- TOC -->
* [Code organization](#code-organization)
  * [Table of Contents](#table-of-contents)
  * [Summary](#summary)
  * [Motivation](#motivation)
    * [Goals](#goals)
    * [Non-Goals](#non-goals)
  * [Proposal](#proposal)
    * [sigs.k8s.io/cluster-api/api](#sigsk8siocluster-apiapi)
    * [sigs.k8s.io/cluster-api/utils](#sigsk8siocluster-apiutils)
    * [sigs.k8s.io/cluster-api](#sigsk8siocluster-api)
    * [sigs.k8s.io/cluster-api/test](#sigsk8siocluster-apitest)
  * [Code organization](#code-organization-1)
    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
    * [Security Model](#security-model)
    * [Risks and Mitigations](#risks-and-mitigations)
      * [Improper assumptions](#improper-assumptions-)
      * [Excessive use of /internal](#excessive-use-of-internal)
  * [Implementation History](#implementation-history)
<!-- TOC -->

## Summary

> This proposal documents principle and guidelines about code organization in Cluster API.

Reviewer: minor nit

```suggestion
This proposal documents the principles and guidelines about code organization in Cluster API.
```

> This is relevant both to maintainer, contributors and to everyone importing Cluster API as go dependencies. 

Reviewer: minor nit

```suggestion
This is relevant both to maintainers, contributors and to everyone importing Cluster API as go dependencies. 
```

Notably, this proposal also clarifies the guarantees that apply to different part of the codebase. 

## Motivation

> Ensuring a clean code organization is an continuous effort in projects like Cluster API.

Reviewer: minor nit

```suggestion
Ensuring a clean code organization is a continuous effort in projects like Cluster API.
```

> However, in this particular moment there are few relevant changes/issue impacting code organization,

Reviewer: minor nit

```suggestion
However, in this particular moment there are few relevant changes/issues impacting code organization,
```

so we decided to write this proposal to ensure a shared and clear understanding of the target end state. 

> Issue about code organization:

Reviewer: minor nit

```suggestion
Issues about code organization:
```

- [Importing API types pulls in lots of dependencies](https://github.com/kubernetes-sigs/cluster-api/issues/9011)
- [Make packages related to in-place updates public](https://github.com/kubernetes-sigs/cluster-api/pull/13507)
- [Expose Machine drain planning as a reusable library function](https://github.com/kubernetes-sigs/cluster-api/issues/13650)

On top of that, the Cluster API codebase is under pressure from a significative increase of the number

> of CVEs in golang and in project direct and indirect go dependencies.

Reviewer: i think we could drop "project" from this sentence.

```suggestion
of CVEs in golang and in direct and indirect go dependencies.
```

or, if we wanted to be more specific we could say:

```suggestion
of CVEs in golang and in direct and indirect go dependencies for the project.
```

Last but not least, the end state should also tackle most of the code organization tech debt, e.g.:

- Complete the cleanup of the `exp` folder 
- Ensure a cleaner separation of the components hosted in the Cluster API code base (core Cluster API, KCP, CABPK)

### Goals

- Provide a clean definition of what part of Cluster API are intended for "general usage" as a go dependency in other projects,
  e.g. use go types for Cluster API v1beta2 resources 
  - Document the guarantees that each consumer can rely-on for the code that is intended for "general usage" as a go dependency in other projects.
- Make it possible for consumer to have a "tighter integration" with Cluster API, like e.g. re-use some of the code in CAPI controllers.
  - Document the guarantees that each consumer can rely-on when they opt in to a "tighter integration" with Cluster API.
- Make the Cluster API code organization as simpler, cleaner and maintainable as possible.

### Non-Goals

- Split the Cluster API repository in multiple repositories. 

>   - We already experimented this approach in the past, and the additional toil was not sustainable for 

Reviewer: minor nit

```suggestion
  - We already experimented with this approach in the past, and the additional toil was not sustainable for 
```

rest of the codebase, but in this case we will implement this separation by introducing dedicated nested go modules, 
more specifically `sigs.k8s.io/cluster-api/api`, `sigs.k8s.io/cluster-api/utils`.

The `sigs.k8s.io/cluster-api/api` and the `sigs.k8s.io/cluster-api/utils` go modules are expected to cover most of the needs
of other projects consuming Cluster API as a go dependency, and those go modules will have the highest level of guarantee 
the Cluster API project can offer.

If we continue the parallel with Kubernetes, at this point it is worth to notice that the main Kubernetes repository,
which corresponds to the `k8s.io/kubernetes` go module, is not intended to be used as a go library by other projects (use at your own risk).

In Cluster API instead we are going to make an additional effort to provide a limited set of guarantees 
for both the "top level" `sigs.k8s.io/cluster-api` go module and the `sigs.k8s.io/cluster-api/test` go module.

While we understand that "limited set of guarantees" might seem not ideal, we kindly invite you to carefully
consider the other side of the coin.

> - The Cluster API maintainer team, is willing to go above and behind to help projects willing to have a tighter 

Reviewer: minor nit

```suggestion
- The Cluster API maintainer team, is willing to go above and beyond to help projects willing to have a tighter 
```

  integration with this project, and we are making it possible with a clear contract/well defined trade-offs 
  documented in the following paragraphs.

> - There is  clear path for consumers willing to achieve more guarantees on the code they are relying on, which is

Reviewer: minor nit

```suggestion
- There is a clear path for consumers willing to achieve more guarantees on the code they are relying on, which is
```

  to engage with maintainers to discuss moving code from `sigs.k8s.io/cluster-api` / `sigs.k8s.io/cluster-api/test` to
  `sigs.k8s.io/cluster-api/utils`.

> - Even if the `sigs.k8s.io/cluster-api` and the `sigs.k8s.io/cluster-api/test` go modules, offer limited guarantees,

Reviewer: minor nit

```suggestion
- Even if the `sigs.k8s.io/cluster-api` and the `sigs.k8s.io/cluster-api/test` go modules offer limited guarantees,
```

| `sigs.k8s.io/cluster-api/test`  | No                          | Allowed                                                                    | If necessary                                                               | Best effort                               | Allowed                                  | If required to fix CVE                   |

### sigs.k8s.io/cluster-api/api

The `sigs.k8s.io/cluster-api/api` go module is defined in the `/api `folder, and it will contain go types
for each API resource defined in Cluster API. For sake of simplicity, also KCP and CABPK are included.

This is the go module that will be most widely used by Cluster API consumers, and as a consequence it is where we 
will offer the wider set of guaranteed to consumers. More specifically, most of the types defined in this module 
have to comply both to Kubernetes API guarantees and Kubernetes semver guarantees.

> Kubernetes API guarantees are documents in [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/),

Reviewer: minor nit

```suggestion
Kubernetes API guarantees are documents in the [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/),
```

shared by core Cluster API, Cluster API providers and systems built on top of Cluster API. It is not, and it should not
become a general purpose golang library (providing general purpose golang library is not in the scope of the Cluster API project).

Accordingly, moving code to `sigs.k8s.io/cluster-api/utils` will be subject to a careful evaluation from maintainers;
the decision will also consider factors like:
- Community interest / number of possible consumers / number of candidate maintainers 
- The readiness of the code to become part of a library intended for broader usage across the CAPI ecosystem 
- The notion of [the complexity budget](https://cluster-api.sigs.k8s.io/user/manifesto#the-complexity-budget)
- etc.

Last but not least, please note that the `sigs.k8s.io/cluster-api` go module it is also allowed bump to the min go version required to build
this package also on release branches, e.g. for fixing CVEs in this module or in one of its dependencies.

### sigs.k8s.io/cluster-api/test

The `sigs.k8s.io/cluster-api/test` go module is defined in the `/test` folder.

> This go module has a few function:

Reviewer: minor nit

```suggestion
This go module has a few functions:
```

* `/bootstrap/kubeadm`: CABPK provider, internal code/not for general consumption, limited guarantee
* `/controlplane/kubeadm`: KCP internal, internal code/not for general consumption, limited guarantee
* `/pkg`: internal code shared among the above components/not for general consumption, limited guarantee
* `/test`: public e2e test code & test providers, limited guarantee (`sigs.k8s.io/cluster-api/test` go module)
* `/hack`: scripting, CI, no guarantees (go module: hack/tools)
* `/docs`: documentation
* `/CHANGELOG`: documentation

Please note we are also going to ensure a consistent struct for all the core components (core Cluster API, KCP, CABPK),
having the same set of folders in `/core`, `/bootstrap/kubeadm`, `/controlplane/kubeadm`:
* `/config`
* `/controllers/<reconciled type>/`
* `/pkg/{setup,...}`
* `/webhooks/{conversion,...}`
* `/main.go`

### Implementation Details/Notes/Constraints

Cluster API already have a `sigs.k8s.io/cluster-api/test` go module since a long time and the
the `sigs.k8s.io/cluster-api/api` go module has been recently introduced. 

Other changes described in this proposal, are already ongoing since a few release and have been already

> discussed in the office hours. E.g. cleaning up the `/exp` folder.

Reviewer: minor nit

```suggestion
Other changes described in this proposal are already ongoing for a few releases and have been discussed in the office hours. E.g. cleaning up the `/exp` folder.
```

The remaining changes, like the introduction of the `sigs.k8s.io/cluster-api/utils` go module 

> and implementing code organization changes will be implemented incrementally.

Reviewer: minor nit

```suggestion
and code organization changes will be implemented incrementally.
```
